### PR TITLE
web-client: Enable HD wallet key derivation

### PR DESCRIPTION
**Pull request by sisou**
Add classes/structs and methods to support HD wallet key derivation as was possible with the core-js library.

- [ ] Review of encryption/decryption methods by @paberr and/or @hrxi for best practices
- [x] Move of CryptoUtils/hashing methods into their own worker so they don't block the main thread

### CODE REVIEW

File: key-derivation/src/lib.rs [hunk 1]
```diff
@@ -5,8 +5,9 @@ use nimiq_hash::{hmac::*, sha512::Sha512Hash};
 use nimiq_keys::{Address, Ed25519PublicKey, PrivateKey};
 use nimiq_serde::Serialize;
 use regex::Regex;
+use serde::Deserialize;
```
Comment (Eligioo) [inline key-derivation/src/lib.rs]: Instead of using `serde::Deserialize`, shouldn't we use `nimiq_serde::Deserialize`?

File: mnemonic/src/lib.rs [hunk 2]
```diff
@@ -58,6 +63,106 @@ impl Entropy {
...
+    pub fn from_encrypted(buf: &[u8], key: &[u8]) -> Result<Entropy, String> {
+        let version = buf[0];
+        let rounds_log = buf[1];
+        if rounds_log > 32 {
+            return Err(format!("Rounds out-of-bounds: 2^{}", rounds_log));
+        }
+        let rounds = 2u32.pow(rounds_log as u32);
+
+        match version {
+            1 => unimplemented!(),
```
Comment (Eligioo) [inline mnemonic/src/lib.rs:127]: Are we going to fully discard older versions?

File: web-client/example/README.md [hunk 3]
```diff
@@ -24,7 +24,7 @@ to build the WASM files. Then go into `example/node` and run
 yarn
 ```
 
-to install the dependencies. Other package managers will also work, but the `preinstall` script uses `yarn`.
+to install the dependencies. Other package managers will also work.
```
Comment (Eligioo) [inline web-client/example/README.md:27]: Yarn isn't mandatory all together anymore?

**Review by Eligioo — COMMENTED** [review]
Halfway through the review, already some comments.

File: web-client/example/README.md (hunk 3 above)
Comment (sisou) [reply web-client/example/README.md:27]: Not here, as I removed the preinstall script (below). So packages can be installed with any manager for the example.

File: web-client/src/primitives/es256_signature.rs [hunk 4]
```diff
@@ -17,24 +17,22 @@ impl ES256Signature {
     /// Deserializes an ES256 signature from a byte array.
     ///
     /// Throws when the byte array contains less than 64 bytes.
-    #[wasm_bindgen(js_name = fromBytes)]
-    pub fn deserialize(bytes: &[u8]) -> Result<ES256Signature, JsError> {
+    pub fn unserialize(bytes: &[u8]) -> Result<ES256Signature, JsError> {
```
Comment (Eligioo) [inline web-client/src/primitives/es256_signature.rs]: I assume you switched from `deserialize` to `unserialize` to stay closer to the core-js PoW naming convention? 

Though throughout the rest of the Rust codebase, `deserialize` is the naming convention for deserializing bytes. And as the `web-client` is breaking anyway compared to the PoW core-js, imo this is the right time to use a different naming convention. Then I would be in favor of using `deserialize` over `unserialize`.

File: web-client/extras/launcher/cryptoutils-worker-proxy.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import type { Remote } from 'comlink';
+
+interface CryptoUtilsWorker {
+    otpKdf(message: Uint8Array, key: Uint8Array, salt: Uint8Array, iterations: number): Promise<Uint8Array>;
+}
+
+let remote: Remote<CryptoUtilsWorker> | undefined;
+
+// class CryptoUtilsProxy {}
```
Comment (Eligioo) [inline web-client/extras/launcher/cryptoutils-worker-proxy.ts]: Forgot to remove?

File: web-client/extras/lib/Secret.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
+import { CryptoUtils, Hash, PrivateKey } from '@nimiq/core';
+import { BufferUtils } from './BufferUtils';
+import { Entropy } from './Entropy';
+import { SerialBuffer } from './SerialBuffer';
+import { Serializable } from './Serializable';
+
+abstract class Secret extends Serializable {
```
Comment (Eligioo) [inline web-client/extras/lib/Secret.ts:7]: Would it be an idea to also port over the tests? For the `ExtendedPrivateKey`, `Secret`, etc.?

File: mnemonic/src/lib.rs (hunk 2 above)
Comment (sisou) [reply mnemonic/src/lib.rs:127]: The older versions use a KDF OTP function that core-rs-albatross no longer implements.

File: web-client/extras/lib/Secret.ts (hunk 6 above)
Comment (sisou) [reply web-client/extras/lib/Secret.ts:7]: I can try, but then we need a JS test-runner, too :/

Comment (sisou) [reply web-client/extras/lib/Secret.ts:7]: Added the relevant tests with `vitest` :muscle:

**Comment by sisou** [comment]
@jsdanielh Rebased and squashed a few earlier commit together.

**Comment by Eligioo** [comment]
Has this been further reviewed?

**Pull request merged**